dashboard/app/reporting_email.go

```diff
@@ -102,6 +106,113 @@ func (cfg *EmailConfig) Validate() error {
 	return nil
 }
 
+// handleCoverageReports sends reports for the previous period.
+// Assuming it is called June 15, the monthly report will cover April-May diff.
+func handleCoverageReports(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	curHostPort := r.URL.Host
+	targetDate := civil.DateOf(timeNow(ctx)).AddMonths(-1)
+	periods, err := coveragedb.GenNPeriodsTill(2, targetDate, "month")
+	if err != nil {
+		msg := fmt.Sprintf("error generating coverage report: %s", err.Error())
+		log.Errorf(ctx, "%s", msg)
+		http.Error(w, "%s: %w", http.StatusBadRequest)
+		return
+	}
+	wg := sync.WaitGroup{}
+	for nsName, nsConfig := range getConfig(ctx).Namespaces {
+		emailTo := nsConfig.Coverage.EmailRegressionsTo
+		minDrop := nsConfig.Coverage.RegressionThreshold
+		if emailTo != "" {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if err := sendNsCoverageReport(ctx, nsName, emailTo, curHostPort, periods, minDrop); err != nil {
+					msg := fmt.Sprintf("error generating coverage report for ns '%s': %s", nsName, err.Error())
+					log.Errorf(ctx, "%s", msg)
+					w.Write([]byte(msg + "\n"))
+					return
+				}
+				w.Write([]byte(fmt.Sprintf("sent ns '%s' report to '%s'\n", nsName, emailTo)))
+			}()
+		}
+	}
+	wg.Wait()
+}
+
+func sendNsCoverageReport(ctx context.Context, ns, email, domain string,
+	period []coveragedb.TimePeriod, minDrop int) error {
+	var days int
+	for _, p := range period {
+		days += p.Days
+	}
+	periodFrom := fmt.Sprintf("%s %d", period[0].DateTo.Month.String(), period[0].DateTo.Year)
+	periodTo := fmt.Sprintf("%s %d", period[1].DateTo.Month.String(), period[1].DateTo.Year)
+	table, err := coverageTable(ctx, ns, period, minDrop)
+	if err != nil {
+		return fmt.Errorf("coverageTable: %w", err)
+	}
+	args := struct {
+		Namespace      string
+		PeriodFrom     string
+		PeriodFromDays int
+		PeriodTo       string
+		PeriodToDays   int
+		Link           string
+		Table          string
+	}{
+		Namespace:      ns,
+		PeriodFrom:     periodFrom,
+		PeriodFromDays: period[0].Days,
+		PeriodTo:       periodTo,
+		PeriodToDays:   period[1].Days,
+		Link: fmt.Sprintf("https://%s/%s/coverage"+
```

Comment:
This link will become broken soon. Is there a way to test it? Or at least use some common code to generate the link?

Reply:
This is the point where cost of testing looks high... But agree, scaling the feature usage we can get the problem.
I'll think how to do it.
Thanks!

Reply:
Added function.